### PR TITLE
[Hexagon] Add Hexagon Load Widening Pass

### PR DESCRIPTION
**Pull request by yandalur**
Extend existing store widening pass to widen load instructions.

This patch also borrows the alias check algorithm from AMDGPU's load store widening pass.

Widened load instruction is inserted before the first candidate load instruction.
Widened store instruction is inserted after the last candidate store instruction.
This method helps avoid moving uses/defs when replacing load/store instructions with their widened equivalents.

The pass has also been extended to
* Generate 64-bit widened stores
* Handle 32-bit post increment load/store
* Handle stores of non-immediate values
* Handle stores where the offset is a GlobalValue

### CODE REVIEW

**Comment by yandalur** [comment]
@iajbar @androm3da @kparzysz Could you please review this PR?

File: llvm/lib/Target/Hexagon/HexagonLoadStoreWidening.cpp [hunk 1]
```diff
@@ -0,0 +1,915 @@
...
+    return false;
+  for (const auto &BOp : B->operands()) {
+    if (!BOp.isReg())
+      continue;
+    if ((BOp.isDef() || BOp.readsReg()) && ARegDefs.contains(BOp.getReg()))
+      return false;
+    if (BOp.isDef() && ARegUses.contains(BOp.getReg()))
+      return false;
+  }
+  return true;
```
Comment (androm3da) [inline llvm/lib/Target/Hexagon/HexagonLoadStoreWidening.cpp:315]: Are there any ordering dependencies here that should be checked with `LLVM_ENABLE_REVERSE_ITERATION` enabled?

Comment (yandalur) [reply llvm/lib/Target/Hexagon/HexagonLoadStoreWidening.cpp:315]: This loop doesn't depend on the iteration order. I do not see any other loop that might depend on non deterministic iteration order. I tested a build with LLVM_REVERSE_ITERATION:BOOL=ON and I did not see any issues.

**Comment by androm3da** [comment]
LGTM but someone else should probably approve this.

**Review by iajbar — APPROVED** [review]

**Comment by androm3da** [comment]
> LLVM Buildbot has detected a new failure on builder `clang-x64-windows-msvc` running on `windows-gcebot2` while building `llvm` at step 4 "annotate".

Hopefully this would be fixed by #117414

**Pull request merged**